### ISSUE TITLE
array_merge(): Argument #1 is not an array

### ISSUE THREAD

**Issue by tqisjim** [rydurham/Sentinel#190]
The following error was thrown:

array_merge(): Argument #1 is not an array

Partial stack trace:
vendor/rydurham/sentinel/src/Sentinel/Traits/SentinelRedirectionTrait.php (line 87)
vendor/rydurham/sentinel/src/Sentinel/Traits/SentinelRedirectionTrait.php (line 30)
vendor/rydurham/sentinel/src/controllers/RegistrationController.php (line 81)

The argument throwing the error, "$payload", is an instance of 
Sentinel\Controllers\RegistrationController

This problem seems to be a bug that never got tested.

As a separate design issue, I am submitting my registration via JSON, so the redirect()->back() response is inappropriate.

*update: the payload argument is actually null, based on the default setting in Sentinel\DataTransferObjects\ExceptionResponse*

**Comment by rydurham** [comment]
Did this occur in a post request to the registration route?  Which version of Sentinel are you using?

**Comment by tqisjim** [comment]
No.  I'm using Sentinel with AngularJS.  
Version 2.2.

**Comment by tqisjim** [comment]
I am attaching a patch.

[SentinelRedirectionTrait_php.txt](https://github.com/rydurham/Sentinel/files/55863/SentinelRedirectionTrait_php.txt)

**Comment by tqisjim** [comment]
I seem to have encountered more problems based on my interpretation of the config file:

    /*
    |--------------------------------------------------------------------------
    | URL Redirection for Method Completion
    |--------------------------------------------------------------------------
    |
    | Upon completion of their tasks, controller methods will look-up their
    | return destination here. You can specify a route, action or URL.
    | If no action is specified a JSON response will be returned.
    |
    */

If no action is specified, the following error is thrown:

  [ErrorException]
  call_user_func_array() expects parameter 1 to be a valid callback, no array
   or string given

This error is thrown in line 105 of SentinelRedirectionTrait.php if either the "route" or "action" specification is missing.

To fix this problem, change lines 100-102 as follows:

        if (is_null($key)) {
            return null;
        }

I am attaching another patch.

[SentinelRedirectionTrait_php.txt](https://github.com/rydurham/Sentinel/files/55951/SentinelRedirectionTrait_php.txt)

**Comment by tqisjim** [comment]
Encountered one more problem:

    [Illuminate\Database\QueryException]
    SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'pari
    stweed' for key 'users_username_unique' (SQL: insert into `users` (`usernam
    e`, `email`, `password`, `updated_at`, `created_at`) values (paristweed, pa
    ristweed@tqis.com, y$.XHbXrQVO0zCdEUHrzuS8OcNNHSYvRkUQjLWQUfJqPiKokygoX0oG,
    2015-12-09 16:22:36, 2015-12-09 16:22:36))

This error is not caught in the try{} block (lines 48-113) in SentryUserRepository.php

This error is thrown on a duplicate username.  Sentry's validation only checks for duplicate emails.  So the *UserExistsException* is not thrown, and the exception propogates downstream. The use case seems fairly common:  Separate users attempt to register using the same username, eg *jim*.

Since the username function is a sentinel enhancement, validation should be written into your code.  At least, the issue can be resolved by broadening the catch criteria in line 113.

**Comment by tqisjim** [comment]
One final comment involving *SentryUserRepository.php* starting at line 46:

    public function store($data)
    {
        try {
            // ...            

            // Return a response
            return new SuccessResponse($message, $payload);
        } catch (UserExistsException $e) {
            $message = trans('Sentinel::users.exists');

            return new ExceptionResponse($message);
        }
    }

Because this method as **public**, the *model* is decoupled from the *controller* constraints defined in RegistrationController.php.  Unfortunately, no one can take advantage of the architecture because the results are encapsulated in implementations of your BaseResponse class, where all the values are **protected**.  Your approach seems inconsistent.

There's not much point in calling public methods where the returned results are inaccessible.

**Comment by rydurham** [comment]
I disagree.  The [BaseResponse](https://github.com/rydurham/Sentinel/blob/master/src/Sentinel/DataTransferObjects/BaseResponse.php) object is intended only to be a DTO.   All of the contents of that object can be retrieved via public getter methods when necessary.

**Comment by rydurham** [comment]
I have started looking into the various problems you have mentioned here.   First things first:  make sure you have the "X-Requested-With" header enabled on the ajax calls you make, this is how Laravel detects an ajax call.  After that, you need to explicitly disable the views in the Sentinel config file: ```'views_enabled' =>  false,```.    This is not ideal, but that is how it works for now. 

With these settings in place, I am able to hit all of the default routes with Postman on my dev box and no errors are triggered.   Can you tell me the exact URLs you are hitting and the parameters you are sending that trigger the problem(s)?  Feel free to  put your sentinel config file in a pastebin and send that to me as well. 

As for usernames,  the Registration controller that comes with the package checks the uniqueness of the requested username using a [Form Request](https://github.com/rydurham/Sentinel/blob/master/src/Sentinel/FormRequests/RegisterRequest.php) before handing the data off to the User Repository.  If you are bypassing the default controller and going straight to the Repository, you should implement the uniqueness logic yourself.

**Comment by tqisjim** [comment]
I stand corrected on the accessor.   couldn't access the method isSuccess() and jumped to the wrong conclusion.

**Comment by tqisjim** [comment]
I acknowledge my mistake in the last post.  And I made a suggestion about how to handle ajax/pjax.  But in the remaining issues, sentinel crashes, and my server throws a 500 response.  I always consider fatal exceptions as bugs, even if there's a work-around.  

I cut and pasted the specific error responses in my earlier posts.  In summary:

1.  ExceptionResponse.php is always going to crash in the described case because *payload* is defined as null, and you explicitly pass that null value to array_merge(),  SentinelRedirectionTrait.php (line 87)
2.  Clearly you intend generateUrl() to return null if the redirect action is undefined.  Instead, sentinel crashes because of a null argument exception in SentinelRedirectionTrait.php (line 123).  The problem is the conditional in line 119 doesn't actually check the first argument passed to call_user_func_array().
3.  In SentryUserRepository.php, line 113, you are only checking for the exceptions you expect.  Other uncaught errors cause sentinel to crash.

These are code mistakes.  Acknowledge them, own them, and fix them.  I've already provided enough specific details rather than expecting you to replicate.  Although you probably need to define more unit tests, since these were never detected.

We apparently agree that the functionality described in config/sentinel.php (line 117) isn't fully implemented.  I haven't tested thoroughly, but the patches I submitted should provide the desired functionality.

The "I disagree." comment threw me off.   In fact, we *do* agree.  And pointing out that the accessor methods are public clears the confusion.  With that correction, I am able to get all the functionality I need.

**Comment by rydurham** [comment]
I am not saying this code is perfect.   Mistakes will be found and they will be fixed.   However, I do think that having the ability to replicate an issue is an important part of the debugging process.   All I am asking for is more context.  

Tell me more about the routes you are hitting and the data you are sending to those routes, or submit a PR with failing tests.

**Comment by rydurham** [comment]
Closing for lack of activity.

**Comment by cliffordy** [comment]
Hi rydurham,

Got the same error as posted above. 
ErrorException in SentinelRedirectionTrait.php line 108:
Argument 1 passed to Sentinel\Controllers\UserController::generateUrl() must be of the type array, null given, called in ...\vendor\rydurham\sentinel\src\Sentinel\Traits\SentinelRedirectionTrait.php on line 52 and defined

It happens when i register a new user. Please help!

**Comment by rydurham** [comment]
Are you using the default sentinel config values or have you made any customizations?

**Comment by cliffordy** [comment]
Am using the defaults of sentinel. The data saves correctly but the redirect is what brings up the error.

**Comment by tqisjim** [comment]
As far as I understand, non-default configuration options are untested and 
unsupported.  There are several known bugs in this code, which doesn't 
seem to have had any formal testing in its development.

Based on my conversation with the author, he may be open about 
contributors who can help write test scripts.  While worthwhile, I 
declilned the opportunity.

I've concluded that, at best, this code is still at the beta stage of 
development.

  -Jim

**Comment by rydurham** [comment]
@cliffordy This should now be fixed in v2.3.5.  However, if you are starting a new Laravel 5.2 project, I recommend you use this package instead: [srlabs/centaur](https://github.com/SRLabs/Centaur).  

@tqisjim I appreciate your help with this.   Communicating via text leaves a lot of room for misunderstanding subtext and intent.  It has been very difficult for me to read your messages here and not interpret them as being rude, which is why my responses to you have been so defensive.  It is true that the tests for this package leave a lot to be desired.  If I had more time available to work on this, I am sure they would be better, but I don't and they are what they are.   I don't appreciate the feeling that you have been offering help with one hand and attacking/insulting with the other.
